File: Top-Interview-Questions-Leetcode/C++/Easy-Collections/Rotate-Array.cpp
#include <iostream>
#include <vector>

// Change the define to test different implementations
#define ROTATE2

class Solution {
private:
    
    void swap(int& a, int& b) {
        a ^= b;
        b ^= a;
        a ^= b;
    }

    void reverse(std::vector<int>& nums, int l, int h) {
        while (l < h) {
            swap(nums[l++], nums[h--]);
        }
    }
    
public:
    void rotate(std::vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        
        if (k == 0) return;
        
        
        // O(n) space complexity
        #ifdef ROTATE
        
        int idx = n - k;
        vector<int> temp;
        
        for (; idx < n; ++idx) temp.push_back(nums[idx]);
        
        for (int i = 0; i < n - k; ++i) temp.push_back(nums[i]);
        
        for (int i = 0; i < n; ++i) nums[i] = temp[i];
        
        #endif
        
        
        // O(k) space complexity
        #ifdef ROTATE1
        
        int idx = n - k;
        vector<int> temp;
        
        for (; idx < n; ++idx) temp.push_back(nums[idx]);
        
        int h = n - 1, l = n - k - 1;
            
        while (l >= 0) {
            nums[h] = nums[l];
            h--, l--;
        }
        
        while (h >= 0) {
            nums[h] = temp.back();
            temp.pop_back();
            h--;
        }
            
        #endif
        
        
        // O(1) space complexity
        #ifdef ROTATE2
        
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
        
        #endif
    }
};

// Example usage
int main() {
    Solution solution;
    std::vector<int> nums = {1, 2, 3, 4, 5, 6, 7};
    int k = 3;
    
    solution.rotate(nums, k);
    
    for (int num : nums) {
        std::cout << num << " ";
    }
    
    return 0;
}

File: Top-Interview-Questions-Leetcode/Easy-Collections/Array/Rotate-Array.py
from typing import List

class Solution:
    
    def __reverse(self, nums: List[int], low: int, high: int) -> None:
        while low < high:
            nums[low] ^= nums[high]
            nums[high] ^= nums[low]
            nums[low] ^= nums[high]
            low += 1
            high -= 1
            
    def _rotate_O_n(self, nums: List[int], k: int) -> None:
        n = len(nums)
        
        first_part = nums[n - k:]
        second_part = nums[:n - k]
        
        for i in range(k):
            nums[i] = first_part[i]
        
        for i in range(k, n):
            nums[i] = second_part[i - k]
    
    def _rotate_O_k(self, nums: List[int], k: int) -> None:
        n = len(nums)
        
        temp = nums[(n - k):]
        
        l = n - k - 1
        h = n - 1
        
        while l >= 0:
            nums[h] = nums[l]
            h -= 1
            l -= 1
        
        for i in range(k):
            nums[i] = temp[i]
    
    def _rotate_O_1(self, nums: List[int], k: int) -> None:
        n = len(nums)
        
        self.__reverse(nums, 0, n - 1)
        self.__reverse(nums, 0, k - 1)
        self.__reverse(nums, k, n - 1)
    
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k %= n
        
        if k == 0:
            return
        
        # Change the method here to test different approaches
        self._rotate_O_1(nums, k)

# Example usage
def main():
    nums = [1, 2, 3, 4, 5, 6, 7]
    k = 3
    solution = Solution()
    solution.rotate(nums, k)
    print(nums)  # Output: [5, 6, 7, 1, 2, 3, 4]

if __name__ == "__main__":
    main()
        